Dynamic Dashboard: Coupons card

selanthiraiyan opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #12716 
Closes: #12728
Part of: #12698

<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Completes the UI and logic for the coupons card in dashboard.

Changes
- Action and remote method for loading coupons using coupon IDs.
- Fetch coupon report and coupons from remote and display those.
- Empty view and content unavailable view.
- Unit tests.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

- Enable the feature flag `dynamicDashboardM2` and build the app.
- Enable the Coupon card on the dashboard.
- Confirm that the card now displays the 3 most active coupons for the selected time range. 
- Validate that the UI matches the design.
- Change the time range and validate the coupons are listed based on the selected time range.
- If there are no coupons available for a selected time range an empty view with the message "No coupon usage during this period" should be displayed.
- Tap on any coupon and validate that the coupon details view is displayed.
- Try editing or deleting the coupon and ensure that it works as expected. 
- Navigate back to the dashboard. Tap View all coupons and confirm that the coupons list view is displayed.
- Disconnect the Internet connect and reload the card. Confirm that the error state is displayed.
- Disable "WooCommerce Analytics" toggle from your wp-admin page. Reload the card and ensure that error view is displayed asking you to check WooCommerce version/settings or contact support.
- Tap the ellipsis button and dismiss the card. Confirm that event `dynamic_dashboard_hide_card_tapped` is tracked with type `coupons`.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

<img src="https://github.com/woocommerce/woocommerce-ios/assets/524475/e6f4ff52-a3b2-4bcd-a7a4-d2b9d9d680b6" width="320"/>


Error cases
| WooCommerce Analytics disabled | No active coupons available for the selected time range | No internet  |
|--------|--------|--------|
| ![image](https://github.com/woocommerce/woocommerce-ios/assets/524475/c3fa650c-a076-447d-8808-0d600800e9f7) | ![image](https://github.com/woocommerce/woocommerce-ios/assets/524475/57aca4dd-7788-4f65-86de-3c78690d46ba) | ![image](https://github.com/woocommerce/woocommerce-ios/assets/524475/fcdcd1ec-e63b-412d-88a6-c5788d527169) | 

wp-admin page "Enable WooCommerce Analytics" toggle

<img width="1230" alt="Screenshot 2024-05-16 at 3 16 36 PM" src="https://github.com/woocommerce/woocommerce-ios/assets/524475/4f47b693-4754-4da3-87c9-ad3aca63625a">


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

selanthiraiyan committed:
Add empty view for coupons card.

selanthiraiyan committed:
Coupons remote - Method to load details of specific coupons based on IDs.

selanthiraiyan committed:
Action to load coupons using coupons IDs.

selanthiraiyan committed:
CouponsRemote - Unit tests for loadCoupons method.

selanthiraiyan committed:
Merge branch 'trunk' into feat/12728-coupons-card

selanthiraiyan committed:
Add tracking event when dismissing coupons card.

selanthiraiyan committed:
Fetch coupons card data from remote.

selanthiraiyan committed:
Show empty and content unavailable views.

selanthiraiyan committed:
Unit tests for MostActiveCouponsCardViewModel.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

selanthiraiyan committed:
Merge branch 'trunk' into feat/12728-coupons-card

selanthiraiyan committed:
Perform appropriate actions upon coupon updation and deletion.

selanthiraiyan committed:
Stop using default values from ProductsRemote.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Coupons/MostActiveCouponsCard.swift:
```diff
@@ -110,6 +124,17 @@ private extension MostActiveCouponsCard {
+            MostActiveCouponsEmptyView()
+
+            Spacer()
+        }
```
Just a suggestion: you can avoid adding `Spacer()` by setting the width of the view to the maximum - that should also push it to the center.

```suggestion
        MostActiveCouponsEmptyView()
            .frame(maxWidth: .infinity)
```

selanthiraiyan replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/12755/commits/5079a5ddf9ad1c8fd39245159d29119a55a6b1d7

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Coupons/MostActiveCouponsCard.swift:
```diff
@@ -159,6 +184,37 @@ private extension MostActiveCouponsCard {
+    }
+
+    var supportForm: some View {
+        NavigationView {
```
Please refrain from using `NavigationView` because it's deprecating soon. `NavigationStack` should be a better alternative.

selanthiraiyan replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/12755/commits/b46ee06380fc126cd18f95b891ce15d8d346b339

itsmeichigo reviewed (approved):
This works great! I noticed a few minor points:

- Should we reorder the items by usage count? In my case, the item with the lower usage count is displayed first, which doesn't make sense when we want to emphasize active items more:

<img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/4ac29e35-bd7f-4692-a232-1ede45764324" width=320 />

Another minor issue: we should keep the divider between the empty state and the view all button too keep them separate IMO:

<img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/0bae6dfe-aa93-4a5a-bc84-fa68e61f6286" width=320 />

Lastly, I wonder if we still fetch items from storage so that we can display the results earlier. The coupon loading request can still be triggered in the background to keep the stored items updated, I just thought if we could display cached items right away that would enhance the UX.

I'm pre-approving the PR so please feel free to merge after addressing the above issues.

selanthiraiyan committed:
Remove Spacers and add a Divider below empty view.

selanthiraiyan committed:
Replace NavigationView with NavigationStack.

selanthiraiyan committed:
Add a view model for MostActiveCouponRow.

selanthiraiyan committed:
Remove Identifiable Conformance as it not needed for Coupon.

selanthiraiyan committed:
Load coupons from storage and refresh in background.

selanthiraiyan committed:
Add unit tests to test that coupons are read from storage as expected.

selanthiraiyan committed:
Merge branch 'trunk' into feat/12728-coupons-card

selanthiraiyan commented:
Thanks for the review and suggestions, @itsmeichigo! 

---

> Should we reorder the items by usage count? In my case, the item with the lower usage count is displayed first, which doesn't make sense when we want to emphasize active items more:

Good catch. The usage count we were showing was the total usage count. I changed that and started displaying the usage count for the selected time range in https://github.com/woocommerce/woocommerce-ios/pull/12755/commits/c17f296fa7a74d97c8a0d83a37ea55079b2ba994. 

> Another minor issue: we should keep the divider between the empty state and the view all button too keep them separate IMO:

Fixed in https://github.com/woocommerce/woocommerce-ios/pull/12755/commits/5079a5ddf9ad1c8fd39245159d29119a55a6b1d7

---

> Lastly, I wonder if we still fetch items from storage so that we can display the results earlier. The coupon loading request can still be triggered in the background to keep the stored items updated, I just thought if we could display cached items right away that would enhance the UX.

Great suggestion. I changed the loading behaviour and started reading coupons from storage in https://github.com/woocommerce/woocommerce-ios/pull/12755/commits/409ef95feccd25fd5d1d946c5f652dbbd2f47ae4.

Please take another look at the PR when you can. 🙇

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Coupons/MostActiveCouponsCard.swift:
```diff
@@ -159,39 +182,36 @@ private extension MostActiveCouponsCard {
-                    .padding(.leading, MostActiveCouponsCard.Layout.padding)
             }
         }
+        .navigationViewStyle(.stack)
```
Nit: this is redundant for `NavigationStack`

selanthiraiyan replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/12755/commits/d8bbf57833d1bbb15db8bacca89182b18b65f705

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Coupons/MostActiveCouponsCardViewModel.swift:
```diff
@@ -66,9 +67,27 @@ final class MostActiveCouponsCardViewModel: ObservableObject {
+
+                /// If all or some coupons are not available in storage
+                /// then fetch from remote
+                if rows.count != couponReports.count {
+                    try await synchronizeCoupons(for: couponReports)
+                } else {
+                    /// Refresh coupons in background
+                    Task { @MainActor in
+                        try await synchronizeCoupons(for: couponReports)
+                    }
+                }
```
I found a possible edge case with my test store. The coupon reports return 3 items and one of them has already been removed so the sync call cannot find that one coupon => the code always fall to the `else` case. I believe this is not a serious case, but that makes me wonder how often this happens and if it makes our storage usage become a waste. I'll leave this for you to decide if we should go ahead without storage (sorry for the back and forth 🙇).

selanthiraiyan replied:
Nice find, Huong! I will keep the storage-related logic as it doesn't create any issues. (Even though it won't be useful for stores that have deleted any of the most active coupons.)

itsmeichigo reviewed (approved):
This still LGTM, but I found an edge case in a comment below.

selanthiraiyan committed:
Merge branch 'trunk' into feat/12728-coupons-card

selanthiraiyan committed:
Remove redundant modifier.
